document POST activities endpoint

Marie000 opened the pull request:
So swagger is doing a good job of documenting single-purpose endpoints, but with our POST /outbox endpoint handling so many different activities, it wasn't really clear. 

As a supplement to the swagger doc, I have created a simple .md file 
Later, I might want to use postman to document the example (which would have the added benefit of being testable, so we would know that the examples do work) but that seems complicated and it does repeat a lot of what swagger already does. 

By doing the documentation, I did notice that it wasn't really clear how creating a document for a publication worked. It was working if you specified the publicationId (the actual id that is in the database) but I fixed it so that using `context: publicationUrl` now works. Which makes more sense. (publicationId still works, but officially the requirement is context)

I also noticed that there are problem with error handling. I will fix that and document it in the same md file. I will also create a link to the activity documentation on the readme and documentation pages.

baldurbjarnason commented on activityDoc.md:
````diff
@@ -0,0 +1,273 @@
+}
+```
+
+### Asssign a Tag to a Publication
````
Typo! 🙂

baldurbjarnason commented:
Having that .md file is a really _really_ useful addition to the swagger-generated docs. I can already tell that I'm going to be using that a lot as a reference 🙂 

And being able to create documents separately from the publication is really useful so it's good that that works. There's a decent chance I would have run into that bug the next time I worked on the import process 👍

Marie000 commented:
I have added more error handling. I meant to do that in another PR but forgot to change branch. 
Some of the things that now return errors:
- assigning a tag to a publication that is not found.
- assigning a tag that is not found to a publication
- deleting a note or publication that is not found or that has already been deleted (do you want to be able to differentiate between something that has been deleted vs. something that never existed?)
- updating a note that is not found

Some checks I have not added yet because I am not sure if they are necessary.
Should these scenarios cause an error?
- creating a document that points to a non-existent publication (invalid publication id)
- creating a document that points to no publication (no publication id)
- creating a note that points to a non-existent document
... well, you get the idea.

baldurbjarnason commented:
Those error situations sound sensible.

> deleting a note or publication that is not found or that has already been deleted (do you want to be able to differentiate between something that has been deleted vs. something that never existed?)

I think getting the same error for not found and already deleted makes sense. So that's fine.

> Should these scenarios cause an error?
> 
> * creating a document that points to a non-existent publication (invalid publication id)
> * creating a document that points to no publication (no publication id)
> * creating a note that points to a non-existent document
>   ... well, you get the idea.

That's a good question. My first instinct is that when these situations happen they are probably caused by errors in the API consumer (i.e. the front end) so having them cause an error would help. That way the front end would unambiguously know that something has gone wrong.

It would help the front end, I think. So, if it's easy to implement and makes sense from the API server's perspective then definitely.

Marie000 commented:
So I have added error handling for scenarios mentioned earlier. (see the .md file for a breakdown)
I also added tests for most of those scenarios. 
As long as Travis doesn't complain about anything. I will merge this. Finally.
